### ISSUE TITLE
VSCode: Changes to dashboard files don't auto-reload Powershell Universal

### ISSUE THREAD

**Issue by JustinGrote** [ironmansoftware/powershell-universal-legacy#66]
#### Description

I expect that when I save a change to a dashboard .ps1 file that Powershell Universal running via the vscode extension will watch that file and restart the dashboard. That is not happening and I have to manually restart the dashboard each time.

#### Version Information

- Universal: 1.4.5
- Universal Hosting Method (IIS, Executable, docker etc): VSCode
- PowerShell: 7.0.3
- Operating System: Windows

#### Internal Issue

We'll fill this out when we create an internal issue in our repository.

**Comment by adamdriscoll** [comment]
We're working on a re-write of the configuration service and associated "watcher". The current implementation is super flaky.

**Comment by JustinGrote** [comment]
@adamdriscoll thanks, I'll write a filesystemwatcher powershellscript in the meantime and will post it here

**Comment by JustinGrote** [comment]
Here is an example watcher I made to restart dashboards when files change. The API key included is not sensitive (my local development key)

TODO: If the dashboard was in debug mode, reconnect to the runspace

```powershell
#requires -module universal

function Watch-PSUProject {
    [CmdletBinding()]
    param(
        #Path of the project to watch
        [ValidateNotNullOrEmpty()][String]$Path = $PWD,
        #Apptoken for your Powershell Universal Dashboard
        [ValidateNotNullOrEmpty()][String]$AppToken,
        #URL of your Powershell Universal Dashboard Instance. Defaults to http://localhost:5000
        [ValidateNotNullOrEmpty()][String]$ComputerName = "http://localhost:5000",
        #Restart all dashboards on any changed file that isn't specifically a dashboard file
        [Switch]$All
    )
    $watcher = [IO.FileSystemWatcher]::new(
        $Path, #Path
        '*.ps1' #Filter
    )
    $watcher.IncludeSubDirectories = $true
    $watcher.EnableRaisingEvents = $true

    $OnChanged = {
        $ErrorActionPreference = 'stop'
        try {
            $ChangedFilePath = $event.SourceEventArgs.FullPath
            Write-Verbose "Change Detected in $ChangedFilePath"
            $lastModified = (get-item $ChangedFilePath).lastwritetime
            if ($lastmodified -le $GLOBAL:__PSUProjectLastModified) {
                Write-Verbose "$ChangedFilePath duplicate refresh action detected for save at $lastModified, ignoring..."
                return
            }
            $GLOBAL:__PSUProjectLastModified = $lastModified

            $AppToken = $event.messagedata.AppToken
            $ComputerName = $event.messagedata.ComputerName
            Connect-PSUServer -ComputerName $ComputerName -apptoken $AppToken

            [UniversalAutomation.Dashboard]$dashboardToRestart = Get-PSUDashboard 
            | Where-Object {(Split-Path $PSItem.FilePath -leaf) -eq (Split-Path $changedFilePath -Leaf)}
            
            if (-not $dashboardToRestart) {
                if (-not $event.MessageData.All) {
                    Write-Verbose "$ChangedFilePath is not a dashboard, no action to take..."
                    return
                }
                Write-Host -fore Cyan "$ChangedFilePath is not a dashboard and -All was specified, restarting all dashboards"
                Get-PSUDashboard | Stop-PSUDashboard
                Get-PSUDashboard | Start-PSUDashboard
                Write-Host -fore Green "All Dashboards Restarted!"
                return
            }

            Write-Host -fore Cyan "$($dashboardToRestart.Name): Change Detected. Restarting..."
            Stop-PSUDashboard -Dashboard $dashboardToRestart
            Start-PSUDashboard -Dashboard $dashboardToRestart
            Write-Host -fore Green "$($dashboardToRestart.Name): Restarted!"
            
        } catch {
            write-host -fore red "ERROR: $PSItem"
        }
    }

    Register-ObjectEvent -InputObject $watcher -EventName Changed -Action $onChanged -MessageData @{
        AppToken     = $AppToken
        ComputerName = $ComputerName
        All          = $All
    }
}
```

**Comment by adamdriscoll** [comment]
Resolved in 1.5.0

**Comment by JustinGrote** [comment]
@adamdriscoll confirmed, it works great and even refreshes the page on changes. Thanks!
